cli/ingress-controller/main.go

```diff
@@ -554,5 +554,19 @@ func registerHandlers(enableProfiling bool, port int, ic *controller.KongControl
 		WriteTimeout:      300 * time.Second,
 		IdleTimeout:       120 * time.Second,
 	}
-	glog.Fatal(server.ListenAndServe())
+	var wg sync.WaitGroup
+	go func() {
+		<-stop
+		wg.Add(1)
```

Comment:
Can this race with `wg.Wait()` on line 571?
What is the reason for not adding first and then blocking?

Reply:
I went back and forth on this one several times. I agree that it looks unusual. My concern is this: It's possible for `(*http.Server).ListenAndServe` to fail immediately, before the goroutine launched at line 558 ever starts running. If `ListenAndServe` fails like that, but we had incremented the `WaitGroup` already, we'll then get stuck waiting on it later. Conversely, if we return from the function without waiting when `ListenAndServer` fails without returning `http.ErrServerClosed`—which it can also do in response to being shut down—then we won't wind up waiting for `(*http.Server).Shutdown` to complete its work. Hence, incrementing the `WaitGroup` only after we've received a signal (or for any other reason that the `stop` channel becomes available) ensures that we only wait on graceful shutdown if we know we're about to initiate that shutdown.

If we make it to line 571 and that goroutine has either not started yet or the `stop` channel remains unavailable, then we haven't started a graceful shutdown, and it's fine to proceed out of the function.

One way to appease your concern is to introduce yet another channel. The goroutine would block on either the `done` or this new channel becoming available. We'd close the new channel here if `(*http.Server).ListenAndServe` exits. I'll post that solution shortly.

Reply:
Please see commit 6cd76eb1ad90e4485d13a768ff584cdd7702e774 and let me know if that's more to your liking.

Reply:
I think this will be a little more robust and avoid the race between the two go routines.

Thanks, Steven!